### ISSUE TITLE
autotools: crosscompilation won't work

### ISSUE THREAD

**Issue by grindhold** [grindhold/libgtkflow#20]
hey @esodan . yesterday i tried to crosscompile libgtkflow as it is very possible that the lib will be used on windows, too.

i tried to run three different compiler configurations:

```
./configure --host=x86_64-w64-mingw32

./configure --host=i686-w64-mingw32

./configure --host=amd64-mingw32msvc
```

But none of them compiles through while native compilation works flawlessly.
Maybe there is something obvious missing that i don't see, because i do not crosscompile stuff on an everyday basis. Do i have to get different glib header files?
I'll dump the outputs for the several configurations here for debugging purposes:

x86_64-w64-mingw32
---------------------------
```
make  all-recursive
make[1]: Entering directory '/home/grindhold/git/libgtkflow'
Making all in libgtkflow
make[2]: Entering directory '/home/grindhold/git/libgtkflow/libgtkflow'
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
  CC       namespace-info.lo
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
  CC       dock.lo
In file included from /usr/include/glib-2.0/gio/gio.h:46:0,
                 from /usr/include/gtk-3.0/gdk/gdkapplaunchcontext.h:28,
                 from /usr/include/gtk-3.0/gdk/gdk.h:32,
                 from /usr/include/gtk-3.0/gtk/gtk.h:30,
                 from dock.c:7:
/usr/include/glib-2.0/gio/gcredentials.h:75:1: error: unknown type name 'uid_t'
 uid_t            g_credentials_get_unix_user      (GCredentials    *credentials,
 ^
/usr/include/glib-2.0/gio/gcredentials.h:79:52: error: unknown type name 'uid_t'
                                                    uid_t           uid,
                                                    ^
In file included from /usr/lib/x86_64-linux-gnu/glib-2.0/include/glibconfig.h:9:0,
                 from /usr/include/glib-2.0/glib/gtypes.h:32,
                 from /usr/include/glib-2.0/glib/galloca.h:32,
                 from /usr/include/glib-2.0/glib.h:30,
                 from dock.c:5:
dock.c: In function 'gtk_flow_dock_get_type':
/usr/include/glib-2.0/glib/gmacros.h:181:53: error: size of array '_GStaticAssertCompileTimeAssertion_0' is negative
 #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                                     ^
/usr/include/glib-2.0/glib/gmacros.h:178:47: note: in definition of macro 'G_PASTE_ARGS'
 #define G_PASTE_ARGS(identifier1,identifier2) identifier1 ## identifier2
                                               ^
/usr/include/glib-2.0/glib/gmacros.h:181:44: note: in expansion of macro 'G_PASTE'
 #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                            ^
/usr/include/glib-2.0/glib/gthread.h:249:5: note: in expansion of macro 'G_STATIC_ASSERT'
     G_STATIC_ASSERT (sizeof *(location) == sizeof (gpointer));       \
     ^
dock.c:496:6: note: in expansion of macro 'g_once_init_enter'
  if (g_once_init_enter (&gtk_flow_dock_type_id__volatile)) {
      ^
In file included from /usr/include/glib-2.0/glib/gasyncqueue.h:32:0,
                 from /usr/include/glib-2.0/glib.h:32,
                 from dock.c:5:
/usr/include/glib-2.0/glib/gthread.h:250:17: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
     (void) (0 ? (gpointer) *(location) : 0);                         \
                 ^
dock.c:496:6: note: in expansion of macro 'g_once_init_enter'
  if (g_once_init_enter (&gtk_flow_dock_type_id__volatile)) {
      ^
In file included from /usr/lib/x86_64-linux-gnu/glib-2.0/include/glibconfig.h:9:0,
                 from /usr/include/glib-2.0/glib/gtypes.h:32,
                 from /usr/include/glib-2.0/glib/galloca.h:32,
                 from /usr/include/glib-2.0/glib.h:30,
                 from dock.c:5:
/usr/include/glib-2.0/glib/gmacros.h:181:53: error: size of array '_GStaticAssertCompileTimeAssertion_1' is negative
 #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                                     ^
/usr/include/glib-2.0/glib/gmacros.h:178:47: note: in definition of macro 'G_PASTE_ARGS'
 #define G_PASTE_ARGS(identifier1,identifier2) identifier1 ## identifier2
                                               ^
/usr/include/glib-2.0/glib/gmacros.h:181:44: note: in expansion of macro 'G_PASTE'
 #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                            ^
/usr/include/glib-2.0/glib/gatomic.h:107:5: note: in expansion of macro 'G_STATIC_ASSERT'
     G_STATIC_ASSERT (sizeof *(atomic) == sizeof (gpointer));                 \
     ^
/usr/include/glib-2.0/glib/gthread.h:251:7: note: in expansion of macro 'g_atomic_pointer_get'
     (!g_atomic_pointer_get (location) &&                             \
       ^
dock.c:496:6: note: in expansion of macro 'g_once_init_enter'
  if (g_once_init_enter (&gtk_flow_dock_type_id__volatile)) {
      ^
/usr/include/glib-2.0/glib/gmacros.h:181:53: error: size of array '_GStaticAssertCompileTimeAssertion_2' is negative
 #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                                     ^
/usr/include/glib-2.0/glib/gmacros.h:178:47: note: in definition of macro 'G_PASTE_ARGS'
 #define G_PASTE_ARGS(identifier1,identifier2) identifier1 ## identifier2
                                               ^
/usr/include/glib-2.0/glib/gmacros.h:181:44: note: in expansion of macro 'G_PASTE'
 #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                            ^
/usr/include/glib-2.0/glib/gthread.h:256:5: note: in expansion of macro 'G_STATIC_ASSERT'
     G_STATIC_ASSERT (sizeof *(location) == sizeof (gpointer));       \
     ^
dock.c:500:3: note: in expansion of macro 'g_once_init_leave'
   g_once_init_leave (&gtk_flow_dock_type_id__volatile, gtk_flow_dock_type_id);
   ^
Makefile:625: recipe for target 'dock.lo' failed
make[2]: *** [dock.lo] Error 1
make[2]: Leaving directory '/home/grindhold/git/libgtkflow/libgtkflow'
Makefile:579: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/grindhold/git/libgtkflow'
Makefile:489: recipe for target 'all' failed
make: *** [all] Error 2
```

i686-w64-mingw32
-------------------------
```
make  all-recursive
make[1]: Entering directory '/home/grindhold/git/libgtkflow'
Making all in libgtkflow
make[2]: Entering directory '/home/grindhold/git/libgtkflow/libgtkflow'
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
  CC       namespace-info.lo
  CC       dock.lo
In file included from /usr/include/glib-2.0/gio/gio.h:46:0,
                 from /usr/include/gtk-3.0/gdk/gdkapplaunchcontext.h:28,
                 from /usr/include/gtk-3.0/gdk/gdk.h:32,
                 from /usr/include/gtk-3.0/gtk/gtk.h:30,
                 from dock.c:7:
/usr/include/glib-2.0/gio/gcredentials.h:75:1: error: unknown type name 'uid_t'
 uid_t            g_credentials_get_unix_user      (GCredentials    *credentials,
 ^
/usr/include/glib-2.0/gio/gcredentials.h:79:52: error: unknown type name 'uid_t'
                                                    uid_t           uid,
                                                    ^
In file included from /usr/include/glib-2.0/glib/gthread.h:32:0,
                 from /usr/include/glib-2.0/glib/gasyncqueue.h:32,
                 from /usr/include/glib-2.0/glib.h:32,
                 from dock.c:5:
dock.c: In function 'gtk_flow_dock_get_type':
/usr/include/glib-2.0/glib/gatomic.h:108:5: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
     (gpointer) __atomic_load_8 ((atomic), __ATOMIC_SEQ_CST);                 \
     ^
/usr/include/glib-2.0/glib/gthread.h:251:7: note: in expansion of macro 'g_atomic_pointer_get'
     (!g_atomic_pointer_get (location) &&                             \
       ^
dock.c:496:6: note: in expansion of macro 'g_once_init_enter'
  if (g_once_init_enter (&gtk_flow_dock_type_id__volatile)) {
      ^
Makefile:625: recipe for target 'dock.lo' failed
make[2]: *** [dock.lo] Error 1
make[2]: Leaving directory '/home/grindhold/git/libgtkflow/libgtkflow'
Makefile:579: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/grindhold/git/libgtkflow'
Makefile:489: recipe for target 'all' failed
make: *** [all] Error 2
```

amd64-mingw32msvc
-------------------------
```
make  all-recursive
make[1]: Entering directory '/home/grindhold/git/libgtkflow'
Making all in libgtkflow
make[2]: Entering directory '/home/grindhold/git/libgtkflow/libgtkflow'
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
  CC       namespace-info.lo
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
  CC       dock.lo
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
  CC       node.lo
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
  CC       nodeview.lo
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
  CC       sink.lo
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
  CC       source.lo
  CCLD     libgtkflow-0.2.la

*** Warning: linker path does not have real file for library -lgtk-3.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libgtk-3 and none of the candidates passed a file format test
*** using a file magic. Last file checked: gtk-update-icon-cache-3.0

*** Warning: linker path does not have real file for library -lgdk-3.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libgdk-3 and none of the candidates passed a file format test
*** using a file magic. Last file checked: /usr/lib/x86_64-linux-gnu/libgdk-3.so.0.1400.5

*** Warning: linker path does not have real file for library -lpangocairo-1.0.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libpangocairo-1.0 and none of the candidates passed a file format test
*** using a file magic. Last file checked: /usr/lib/x86_64-linux-gnu/libpangocairo-1.0.so.0.3600.8

*** Warning: linker path does not have real file for library -lpango-1.0.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libpango-1.0 and none of the candidates passed a file format test
*** using a file magic. Last file checked: /usr/lib/x86_64-linux-gnu/libpango-1.0.so.0.3600.8

*** Warning: linker path does not have real file for library -latk-1.0.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libatk-1.0 and none of the candidates passed a file format test
*** using a file magic. Last file checked: /usr/lib/x86_64-linux-gnu/libatk-1.0.so.0.21609.1

*** Warning: linker path does not have real file for library -lcairo-gobject.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libcairo-gobject and none of the candidates passed a file format test
*** using a file magic. Last file checked: /usr/lib/x86_64-linux-gnu/libcairo-gobject.so.2.11400.2

*** Warning: linker path does not have real file for library -lcairo.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libcairo and none of the candidates passed a file format test
*** using a file magic. Last file checked: /usr/lib/x86_64-linux-gnu/libcairo.so.2.11400.2

*** Warning: linker path does not have real file for library -lgdk_pixbuf-2.0.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libgdk_pixbuf-2.0 and none of the candidates passed a file format test
*** using a file magic. Last file checked: /usr/lib/x86_64-linux-gnu/libgdk_pixbuf-2.0.so.0.3100.4

*** Warning: linker path does not have real file for library -lgio-2.0.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libgio-2.0 and none of the candidates passed a file format test
*** using a file magic. Last file checked: /usr/lib/x86_64-linux-gnu/libgio-2.0.so.0.4400.1

*** Warning: linker path does not have real file for library -lgobject-2.0.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libgobject-2.0 and none of the candidates passed a file format test
*** using a file magic. Last file checked: /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.4400.1

*** Warning: linker path does not have real file for library -lglib-2.0.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libglib-2.0 and none of the candidates passed a file format test
*** using a file magic. Last file checked: /lib/x86_64-linux-gnu/libglib-2.0.so.0.4400.1
*** The inter-library dependencies that have been dropped here will be
*** automatically added whenever a program is linked with this library
*** or is declared to -dlopen it.

*** Since this library must not contain undefined symbols,
*** because either the platform does not support them or
*** it was explicitly requested with -no-undefined,
*** libtool will only create a static version of it.
(echo -e EXPRTS; gcc -E -P  \
- <gtkflow.symbols libgtkflow-0.2.la | sed -e '/^$/d' -e 's/^/ /' -e 's/G_GNUC_[^ ]*//g' \
| sort) > gtkflow.def.tmp && mv gtkflow.def.tmp libgtkflow-0.2.def
gcc: warning: libgtkflow-0.2.la: linker input file unused because linking not done
make  all-am
make[3]: Entering directory '/home/grindhold/git/libgtkflow/libgtkflow'
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
/usr/bin/valac   --library=gtkflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgtkflow --vapi gtkflow-0.2.vapi -H gtkflow.h -C  --gir=GtkFlow-0.2.gir --symbols=gtkflow.symbols namespace-info.vala dock.vala node.vala nodeview.vala sink.vala source.vala
(echo -e EXPRTS; gcc -E -P  \
- <gtkflow.symbols libgtkflow-0.2.la | sed -e '/^$/d' -e 's/^/ /' -e 's/G_GNUC_[^ ]*//g' \
| sort) > gtkflow.def.tmp && mv gtkflow.def.tmp libgtkflow-0.2.def
gcc: warning: libgtkflow-0.2.la: linker input file unused because linking not done
/usr/bin/g-ir-compiler --includedir=. -l `/bin/sed -n "s/^dlname='\([A-Za-z0-9.+-]\+\)'/\1/p" libgtkflow-0.2.la`  GtkFlow-0.2.gir -o GtkFlow-0.2.typelib
no input files
Makefile:998: recipe for target 'GtkFlow-0.2.typelib' failed
make[3]: *** [GtkFlow-0.2.typelib] Error 1
make[3]: Leaving directory '/home/grindhold/git/libgtkflow/libgtkflow'
Makefile:520: recipe for target 'all' failed
make[2]: *** [all] Error 2
make[2]: Leaving directory '/home/grindhold/git/libgtkflow/libgtkflow'
Makefile:579: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/grindhold/git/libgtkflow'
Makefile:489: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by grindhold** [comment]
There is a macro not being set correctly. Everything that uses uid_t is inside a #ifdef G_OS_UNIX #endif directive. I need to figure out a way to assure that this directive is not being defined.

**Comment by grindhold** [comment]
i manually changed the line in my glibconfig.h. at least it leads the i686-w64-mingw32 to build up to the same stage as the amd64-mingw32msvc-build. but this can't be the answer. There must be a more convenient way. I won't fiddle around in my system everytime i want to build a release. Any ideas?

**Comment by esodan** [comment]
Hey sorry for late response, but stop to do your are for last comments please!

You should have set pkg-config configured correctly to find correct pc files to compile against currect platform.

If you plan to crosscompile, you should have an environment correctly setup to point correct header files, as you can see at your compiler try to use /usr/include, this is a path for linux local compile. If you have an environment in a path (like the one on www.gtk.org) you have to set  PKG_CONFIG_PATH to that location and set --host according to that path, this is the way I used to do crosscompile, I see on Federa and Debian you can install mingw64, but you should point to correct path to find all pc files you need to the target host. I have no crosscompile any more, I can't help you more here, but may could in future if you wish.

What I do is to have a VM with windows installed, then install MSYS2 environment, then clone this repository:

https://github.com/Alexpux/MINGW-packages

Then compile and install all dependencies for your software, in this case glib and Gtk+, then you should create your own PKGBUILD file for GFlow and GtkFlow, install them when finish to your system and you're done. (This is more easy to write that to do it)

Once you have a PKGBUILD for each of GFlow and GtkFlow, you can create a pull request to add it to MINGW-packages repository in order to help other to use your software in Windows.

**Comment by grindhold** [comment]
i have a windows machine under my hands right now. how am i supposed to bootstrap mingw-w64-gcc? the package depends on itself.

**Comment by esodan** [comment]
You should install MSYS2 from:

https://msys2.github.io/

on your Windows box, then you'll have a Unix/Linux like terminal where you can use:

makepkg-mingw to build packages

pacman -U to install them

There are more complete information on the above link, you can use to build your software against.

Just take in account that once you get your packages from https://github.com/Alexpux/MINGW-packages you should know that it is a "rolling" repository, then you always get the latest version available with very common breaks.

I'm using GTK+ 3.14.9, in a Windows Box to compile my software, and keep away from the latest 3.16 in there.

**Comment by grindhold** [comment]
okay, i succeeded in compiling gflow, but i fail compiling libgtkflow.
gtkflow will compile but it will not link. it says, that every function exposed by gflow does not exist. gflow resources are all there in the subtree with prefix "/mingw64" i have a headerfile in /mingw64/include/gflow-0.2/libgflow/gflow.h and i have the libraries in /mingw64/lib/ pkg-config --cflags seems to deliver correct values.

**Comment by esodan** [comment]
I need to fix GtkFlow on master to make you can compile on MSYS2/Mingw64.
I'll do it ASAP.

**Comment by grindhold** [comment]
cool. i'm thrilled to see what caused the issue

**Comment by grindhold** [comment]
current status: build with mingw-w64_x86_64 in msys2 fails.
./autogen.sh  runs through flawlessly:

```
 Configuration summary for libgtkflow-0.1.0
   Installation prefix:     /usr/local
   GObject Introspection:   yes
   Documentation:           no
       DevHelp Doc:         no
       Gtk-Doc:             no
       GIR Documented:      no
   Debug:                   no
   Platform:                x86_64-pc-mingw64

Now type `make' to compile.
```

Make however fails:

```
grindhold@windows MINGW64 ~/libgtkflow
$ make
(CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/sh /home/grindhold/libgtkflow/missing autoheader)
rm -f stamp-h1
touch config.h.in
cd . && /bin/sh ./config.status config.h
config.status: creating config.h
config.status: config.h is unchanged
make  all-recursive
make[1]: Verzeichnis „/home/grindhold/libgtkflow“ wird betreten
Making all in libgflow
make[2]: Verzeichnis „/home/grindhold/libgtkflow/libgflow“ wird betreten
/mingw64/bin/valac   --library=gflow-0.2 --vapidir=. --pkg glib-2.0 --pkg gtk+-3.0 --includedir libgflow --vapi gflow-0.2.vapi -H gflow.h -C  --gir=GFlow-0.2.gir --symbols=gflow.symbols namespace-info.vala gflow-dock.vala gflow-node.vala gflow-simple-node.vala gflow-simple-sink.vala gflow-simple-source.vala gflow-sink.vala gflow-source.vala
  CC       namespace-info.lo
  CC       gflow-dock.lo
In file included from C:/msys64/mingw64/lib/glib-2.0/include/glibconfig.h:9:0,
                 from C:/msys64/mingw64/include/glib-2.0/glib/gtypes.h:32,
                 from C:/msys64/mingw64/include/glib-2.0/glib/galloca.h:32,
                 from C:/msys64/mingw64/include/glib-2.0/glib.h:30,
                 from gflow-dock.c:5:
gflow-dock.c: In function 'gflow_dock_get_type':
C:/msys64/mingw64/include/glib-2.0/glib/gmacros.h:181:53: error: size of array '_GStaticAssertCompileTimeAssertion_0' is negative
 #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                                     ^
C:/msys64/mingw64/include/glib-2.0/glib/gmacros.h:178:47: note: in definition of macro 'G_PASTE_ARGS'
 #define G_PASTE_ARGS(identifier1,identifier2) identifier1 ## identifier2
                                               ^
C:/msys64/mingw64/include/glib-2.0/glib/gmacros.h:181:44: note: in expansion of macro 'G_PASTE'
 #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                            ^
C:/msys64/mingw64/include/glib-2.0/glib/gthread.h:249:5: note: in expansion of macro 'G_STATIC_ASSERT'
     G_STATIC_ASSERT (sizeof *(location) == sizeof (gpointer));       \
     ^
gflow-dock.c:416:6: note: in expansion of macro 'g_once_init_enter'
  if (g_once_init_enter (&gflow_dock_type_id__volatile)) {
      ^
In file included from C:/msys64/mingw64/include/glib-2.0/glib/gasyncqueue.h:32:0,
                 from C:/msys64/mingw64/include/glib-2.0/glib.h:32,
                 from gflow-dock.c:5:
C:/msys64/mingw64/include/glib-2.0/glib/gthread.h:250:17: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
     (void) (0 ? (gpointer) *(location) : 0);                         \
                 ^
gflow-dock.c:416:6: note: in expansion of macro 'g_once_init_enter'
  if (g_once_init_enter (&gflow_dock_type_id__volatile)) {
      ^
In file included from C:/msys64/mingw64/lib/glib-2.0/include/glibconfig.h:9:0,
                 from C:/msys64/mingw64/include/glib-2.0/glib/gtypes.h:32,
                 from C:/msys64/mingw64/include/glib-2.0/glib/galloca.h:32,
                 from C:/msys64/mingw64/include/glib-2.0/glib.h:30,
                 from gflow-dock.c:5:
C:/msys64/mingw64/include/glib-2.0/glib/gmacros.h:181:53: error: size of array '_GStaticAssertCompileTimeAssertion_1' is negative
 #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                                     ^
C:/msys64/mingw64/include/glib-2.0/glib/gmacros.h:178:47: note: in definition of macro 'G_PASTE_ARGS'
 #define G_PASTE_ARGS(identifier1,identifier2) identifier1 ## identifier2
                                               ^
C:/msys64/mingw64/include/glib-2.0/glib/gmacros.h:181:44: note: in expansion of macro 'G_PASTE'
 #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                            ^
C:/msys64/mingw64/include/glib-2.0/glib/gatomic.h:107:5: note: in expansion of macro 'G_STATIC_ASSERT'
     G_STATIC_ASSERT (sizeof *(atomic) == sizeof (gpointer));                 \
     ^
C:/msys64/mingw64/include/glib-2.0/glib/gthread.h:251:7: note: in expansion of macro 'g_atomic_pointer_get'
     (!g_atomic_pointer_get (location) &&                             \
       ^
gflow-dock.c:416:6: note: in expansion of macro 'g_once_init_enter'
  if (g_once_init_enter (&gflow_dock_type_id__volatile)) {
      ^
In file included from C:/msys64/mingw64/include/glib-2.0/glib/gthread.h:32:0,
                 from C:/msys64/mingw64/include/glib-2.0/glib/gasyncqueue.h:32,
                 from C:/msys64/mingw64/include/glib-2.0/glib.h:32,
                 from gflow-dock.c:5:
C:/msys64/mingw64/include/glib-2.0/glib/gatomic.h:108:5: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
     (gpointer) __atomic_load_8 ((atomic), __ATOMIC_SEQ_CST);                 \
     ^
C:/msys64/mingw64/include/glib-2.0/glib/gthread.h:251:7: note: in expansion of macro 'g_atomic_pointer_get'
     (!g_atomic_pointer_get (location) &&                             \
       ^
gflow-dock.c:416:6: note: in expansion of macro 'g_once_init_enter'
  if (g_once_init_enter (&gflow_dock_type_id__volatile)) {
      ^
In file included from C:/msys64/mingw64/lib/glib-2.0/include/glibconfig.h:9:0,
                 from C:/msys64/mingw64/include/glib-2.0/glib/gtypes.h:32,
                 from C:/msys64/mingw64/include/glib-2.0/glib/galloca.h:32,
                 from C:/msys64/mingw64/include/glib-2.0/glib.h:30,
                 from gflow-dock.c:5:
C:/msys64/mingw64/include/glib-2.0/glib/gmacros.h:181:53: error: size of array '_GStaticAssertCompileTimeAssertion_2' is negative
 #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                                     ^
C:/msys64/mingw64/include/glib-2.0/glib/gmacros.h:178:47: note: in definition of macro 'G_PASTE_ARGS'
 #define G_PASTE_ARGS(identifier1,identifier2) identifier1 ## identifier2
                                               ^
C:/msys64/mingw64/include/glib-2.0/glib/gmacros.h:181:44: note: in expansion of macro 'G_PASTE'
 #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                            ^
C:/msys64/mingw64/include/glib-2.0/glib/gthread.h:256:5: note: in expansion of macro 'G_STATIC_ASSERT'
     G_STATIC_ASSERT (sizeof *(location) == sizeof (gpointer));       \
     ^
gflow-dock.c:421:3: note: in expansion of macro 'g_once_init_leave'
   g_once_init_leave (&gflow_dock_type_id__volatile, gflow_dock_type_id);
   ^
Makefile:635: die Regel für Ziel „gflow-dock.lo“ scheiterte
make[2]: *** [gflow-dock.lo] Fehler 1
make[2]: Verzeichnis „/home/grindhold/libgtkflow/libgflow“ wird verlassen
Makefile:584: die Regel für Ziel „all-recursive“ scheiterte
make[1]: *** [all-recursive] Fehler 1
make[1]: Verzeichnis „/home/grindhold/libgtkflow“ wird verlassen
Makefile:495: die Regel für Ziel „all“ scheiterte
make: *** [all] Fehler 2

```

**Comment by esodan** [comment]
What is your build environment?

Have you set *PKG_CONFIG_**PATH* to your build environment
{prefix}/lib/pkgconfig location, prefix depends on where your headers are
located.

**Comment by grindhold** [comment]
```
grindhold@windows MINGW64 ~/libgtkflow
$ env | grep PKG_CONFIG_PATH
PKG_CONFIG_PATH=/mingw64/lib/pkgconfig:/mingw64/share/pkgconfig
```
Seems as if it's set correctly. The former directory contains the relevant .pc-files (e.g. cairo.pc, gtk+3.0.pc glib-2.0.pc and so on.

**Comment by esodan** [comment]
For your configure add:

configure --host=x86_64-w64-mingw32 --prefix={prefix to your installation
path}

Make sure to setup your PKG_CONFIG_PATH to locate correct headers, they
should be located depending on Mingw64 installation.

**Comment by grindhold** [comment]
results in the same problems.

**Comment by esodan** [comment]
Describe your build system, provide gtk+-3.0.pc contents you are building
against and including your configure output

**Comment by grindhold** [comment]
will deliver tomorrow. i don't have access to the windows machine right now.

**Comment by grindhold** [comment]
@esodan 

gtk+-3.0.pc

```
prefix=/mingw64
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include
targets=win32 broadway

gtk_binary_version=3.0.0
gtk_host=x86_64-w64-mingw32

Name: GTK+
Description: GTK+ Graphical UI Library
Version: 3.16.4
Requires: gdk-3.0 atk >= 2.15.1 cairo >= 1.14.0 cairo-gobject >= 1.14.0 gdk-pixbuf-2.0 >= 2.30.0 gio-2.0 >= 2.43.4
Requires.private: atk   epoxy >= 1.0
Libs: -L${libdir} -lgtk-3 
Cflags: -I${includedir}/gtk-3.0 -mms-bitfields
```

gdk-3.0.pc

```
prefix=/mingw64
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include
targets=win32 broadway

Name: GDK
Description: GTK+ Drawing Kit
Version: 3.16.4
Requires: pangowin32 pangocairo gdk-pixbuf-2.0 >= 2.30.0 cairo >= 1.14.0 cairo-gobject >= 1.14.0
Requires.private: gio-2.0 >= 2.43.4     cairo-win32 cairo epoxy >= 1.0
Libs: -L${libdir} -lgdk-3  -lgdi32 -limm32 -lshell32 -lole32 -Wl,-luuid -lwinmm -lz
Cflags: -I${includedir}/gtk-3.0 
```

glib-2.0.pc

```
prefix=/mingw64
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

glib_genmarshal=glib-genmarshal
gobject_query=gobject-query
glib_mkenums=glib-mkenums

Name: GLib
Description: C Utility Library
Version: 2.44.1
Requires.private: 
Libs: -L${libdir} -lglib-2.0 -lintl 
Libs.private: -pthread -lws2_32 -lole32 -lwinmm -lshlwapi  -lintl    
Cflags: -I${includedir}/glib-2.0 -I${libdir}/glib-2.0/include -mms-bitfields
```

**Comment by esodan** [comment]
As for responses, you have a Windows box with Mingw64 build system.

In order to build there, you should have a tarball built with make distchek
from your repository.

You should use makepkg-mingw to create installable package, to create it
you should have a PKGBUILD file. You can get example ones from [1].

After a package has been created you must install it using pacman -U

 [1] https://github.com/Alexpux/MINGW-packages/

**Comment by grindhold** [comment]
okay, i will try.
why does it not suffice to simply install all the dependencies e.g. mingw-w64-x86_64-gtk via pacman?

**Comment by esodan** [comment]
pacman will install all dependencies when used -S, but libgtkflow doesn't
have a release yet, you must install its dependencies first, then build and
install it, using pacman.

**Comment by grindhold** [comment]
as visible in the screenshots in README.md crosscompilation now works as expected

**Comment by esodan** [comment]
You need to update README.md because CMake is no longer in use.

May you want to create a Broadway backend example screen shot, this will
compete with actual flow programing WEB frameworks.

**Comment by grindhold** [comment]
Broadway is a nice idea. Maybe i could set up some kind of live-demo for on my server for it.

**Comment by esodan** [comment]
Lets open a new issue for Broadway backend
